internal/provider_framework/attribute_validators.go

```diff
@@ -0,0 +1,368 @@
+	)
+	return
+}
+
+// valueInSliceAttributeValidator checks that an attribute value is
+// indeed one of the given `acceptableValues`.
+//
+// Instances should be created via valueInSlice function.
+type valueInSliceAttributeValidator struct {
+	acceptableValues []attr.Value
+}
+
+// valueInSlice is a helper to instantiate a valueInSliceAttributeValidator.
+func valueInSlice(acceptableValues ...attr.Value) tfsdk.AttributeValidator {
+	return &valueInSliceAttributeValidator{acceptableValues}
+}
+
+var _ tfsdk.AttributeValidator = (*valueInSliceAttributeValidator)(nil)
+
+func (av *valueInSliceAttributeValidator) Description(ctx context.Context) string {
+	return av.MarkdownDescription(ctx)
+}
+
+func (av *valueInSliceAttributeValidator) MarkdownDescription(_ context.Context) string {
+	return fmt.Sprintf("Ensures that the attribute is one of: %q", av.acceptableValues)
+}
+
+func (av *valueInSliceAttributeValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, res *tfsdk.ValidateAttributeResponse) {
+	if req.AttributeConfig.IsNull() || req.AttributeConfig.IsUnknown() {
+		return
+	}
+
+	tflog.Debug(ctx, "Validating attribute value is one of the acceptable ones", map[string]interface{}{
+		"attribute":        attrPathToString(req.AttributePath),
+		"acceptableValues": av.acceptableValues,
+	})
+
+	for _, acceptable := range av.acceptableValues {
+		if req.AttributeConfig.Equal(acceptable) {
+			return
+		}
+	}
+
+	res.Diagnostics.AddAttributeError(
+		req.AttributePath,
+		fmt.Sprintf("Attribute %q is invalid: %q", attrPathToString(req.AttributePath), req.AttributeConfig),
+		fmt.Sprintf("Must be one of %q", av.acceptableValues),
+	)
+}
+
+// requiredWithAttributeValidator checks that a set of *tftypes.AttributePath,
+// including the attribute it's applied to, are set simultaneously.
+// This implements the validation logic declaratively within the tfsdk.Schema.
+//
+// The provided tftypes.AttributePath must be "absolute",
+// and starting with top level attribute names.
+type requiredWithAttributeValidator struct {
+	attrPaths []*tftypes.AttributePath
+}
+
+// requiredWith is a helper to instantiate requiredWithAttributeValidator.
+func requiredWith(attributePaths ...*tftypes.AttributePath) tfsdk.AttributeValidator {
+	return &requiredWithAttributeValidator{attributePaths}
+}
+
+var _ tfsdk.AttributeValidator = (*requiredWithAttributeValidator)(nil)
+
+func (av requiredWithAttributeValidator) Description(ctx context.Context) string {
+	return av.MarkdownDescription(ctx)
+}
+
+func (av requiredWithAttributeValidator) MarkdownDescription(_ context.Context) string {
+	return fmt.Sprintf("Ensure that if an attribute is set, also these are set: %q", av.attrPaths)
+}
+
+func (av requiredWithAttributeValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, res *tfsdk.ValidateAttributeResponse) {
+	tflog.Debug(ctx, "Validating attribute is set together with other required attributes", map[string]interface{}{
```

Comment:
Is a check for `if req.AttributeConfig.IsNull() || req.AttributeConfig.IsUnknown()` required here?
If not, should there be one for https://github.com/hashicorp/terraform-provider-tls/pull/215/files#diff-c32022ba2194630d49dde2e8e91220e983a86076686d51bbc96f3edc6aba6f29R117 ?

Reply:
https://github.com/hashicorp/terraform-provider-tls/pull/215#discussion_r882099448